dot/core/service.go

```diff
@@ -526,17 +536,21 @@ func (s *Service) HandleSubmittedExtrinsic(ext types.Extrinsic) error {
 		return err
 	}
 
-	rt.SetContextStorage(ts)
-
 	externalExt, err := s.buildExternalTransaction(rt, ext)
 	if err != nil {
 		return fmt.Errorf("building external transaction: %w", err)
 	}
 
+	// ValidateTransaction modifies the trie state so we want to snapshot it
+	// so that the original trie state remains unaffected.
+	temporaryState := ts.Snapshot()
+	rt.SetContextStorage(temporaryState)
 	transactionValidity, err := rt.ValidateTransaction(externalExt)
 	if err != nil {
 		return err
 	}
+	// TODO adding this fails polkadotjs tests
+	// rt.SetContextStorage(nil)
```

Comment:
Not sure why, the whole program crashes badly, anyone would know why?

Reply:
maybe tests are wrong? I feel like we should be setting the context yes?

Reply:
Oh yeah definitely, but clearing it (setting it to `nil`) makes it panic ugly (cgo horrible trace)